### PR TITLE
Login

### PR DESCRIPTION
**Pull request by aayush-kumar-codes**

### CODE REVIEW

File: lib/model/user.dart [hunk 1]
```diff
@@ -0,0 +1,30 @@
+class User {
+  String _username;
+  String _password;
+  String token;
+  String message;
+  String userid;
+  User(this._username, this._password);
+
+  User.map(dynamic obj) {
```
Comment (manishiitg) [inline lib/model/user.dart:9]: why is map used. i dont know

File: lib/services/login.dart [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import 'dart:async';
+import 'dart:convert';
+import 'package:shared_preferences/shared_preferences.dart';
+import 'post.dart';
+import '../model/user.dart';
+
+class Login {
+  static final baseUrl = "https://apistaginghr.excellencetechnologies.in";
+  static final loginUrl = baseUrl + "/attendance/API_HR/api.php";
```
Comment (manishiitg) [inline lib/services/login.dart:9]: these should be moved to configuration or env file. there should be something for dart

File: lib/services/login.dart [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      "action": "login",
+      "username": username,
+      "password": password
+    };
+    return _post
+        .post(loginUrl, body: json.encode(data))
+        .then((dynamic res) async {
+      if (res["error"] >= 1) throw new Exception(res["data"]["message"]);
+      if (res["error"] == 0) {
+        SharedPreferences prefs = await SharedPreferences.getInstance();
```
Comment (manishiitg) [inline lib/services/login.dart:25]: all local stroage stuff should be moved to a seperate service like localstorage.dart

File: lib/services/login.dart [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  Post _post = Post();
+  Future<User> login(String username, String password) {
+    //print(username + "" + password);
+    Map data = {
+      "token": apiKey,
+      "action": "login",
+      "username": username,
+      "password": password
+    };
+    return _post
```
Comment (manishiitg) [inline lib/services/login.dart:20]: api call to be moved to a seperate service

File: lib/model/user.dart (hunk 1 above)
Comment (deepakexcel) [reply lib/model/user.dart:9]: map is used to assign json data to User object to make it access easily like user.userid.

File: lib/services/login.dart (hunk 4 above)
Comment (deepakexcel) [reply lib/services/login.dart:20]: api call is already separate and above is called by http.post in post.dart

File: lib/services/login.dart (hunk 2 above)
Comment (deepakexcel) [reply lib/services/login.dart:9]: ok creating app config file

File: lib/services/login.dart (hunk 3 above)
Comment (deepakexcel) [reply lib/services/login.dart:25]: ok creating storage service

File: lib/LoginScreen.dart [hunk 5]
```diff
@@ -20,28 +20,25 @@ class _LoginScreenState extends State<LoginScreen> {
...
-                  ],
-                ),
-                WelcomeScreen(),
-                LoginPage(),
-                GoogleLogin(),
-                LoginBottom(),
-              ],
-            ),
+    return Scaffold(
+      backgroundColor: Color(0xfff8f9fa), // just add oxff before hexa code
```
Comment (manishiitg) [inline lib/LoginScreen.dart:24]: we shouldn't not use hard coded colors or fonts like this. this causes problems in the long run. instead need to use this always https://flutter.dev/docs/cookbook/design/themes

File: lib/screens/loginPage.dart [hunk 6]
```diff
@@ -1,5 +1,7 @@
 import 'package:flutter/material.dart';
```
Comment (manishiitg) [inline lib/screens/loginPage.dart:1]: we have a file called "screens/loginPart.dart" and we have a file called "LoginPage.dart"

this is not correct at all.

can we follow this https://medium.com/flutter-community/flutter-code-organization-de3a4c219149

Comment (manishiitg) [reply lib/screens/loginPage.dart:1]: also all filename should be lower case only with underscore if needed.

File: lib/services/storage_service.dart [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  StorageUtil._();
+
+  Future _init() async {
+    _preferences = await SharedPreferences.getInstance();
+  }
+
+  // get string
+  static String getString(String key, {String defValue = ''}) {
```
Comment (manishiitg) [inline lib/services/storage_service.dart:25]: no.  the store service won't be like this this. i.e getString putString. etc. this is not of use.

the function need to be like loginUser  logoutUser etc
i.e the specific case we are using storage for

File: lib/bloc/authentication/authentication_bloc.dart [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (event is UserLoggedOut) {
+      yield* _mapUserLoggedOutToState(event);
+    }
+  }
+
+  Stream<AuthenticationState> _mapAppLoadedToState(AppLoaded event) async* {
+    yield AuthenticationLoading(); // to display splash screen
+    try {
+      await Future.delayed(Duration(milliseconds: 500)); // a simulated delay
```
Comment (manishiitg) [inline lib/bloc/authentication/authentication_bloc.dart:37]: why this?

File: lib/services/authentication_services.dart [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import '../model/user.dart';
+import '../services/login.dart';
+
+abstract class AuthenticationService {
+  Future<User> getCurrentUser();
+  Future<User> signInWithEmailAndPassword(String email, String password);
+  Future<void> signOut();
+}
+
+class FakeAuthenticationService extends AuthenticationService {
```
Comment (manishiitg) [inline lib/services/authentication_services.dart:10]: i dont know what is is fake? we dont have any fake

File: lib/services/authentication_services.dart [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  @override
+  Future<User> getCurrentUser() async {
+    return null; // return null for now
+  }
+
+  @override
+  Future<User> signInWithEmailAndPassword(String email, String password) async {
+    final Login api = Login();
+    await api.login(email, password).then((User user) {
+      print(user.token);
```
Comment (manishiitg) [inline lib/services/authentication_services.dart:20]: ??

File: lib/services/authentication_services.dart [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    });
+
+    return User(email, password, user: 'test');
```
Comment (manishiitg) [inline lib/services/authentication_services.dart:23]: this doesn't look like its done properly

File: lib/screens/login_screen.dart [hunk 12]
```diff
@@ -1,42 +1,85 @@
 import 'package:flutter/material.dart';
-import 'package:flutter_excellence_hr/resources/app_colors.dart';
-import '../screens/login_bottom.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
```
Comment (manishiitg) [inline lib/screens/login_screen.dart:2]: we have two screens? login screen and login_page. i dont understand why two screens for login?

File: lib/model/user.dart [hunk 13]
```diff
@@ -5,7 +5,7 @@ class User {
   String message;
   String userid;
   int error;
-  User(this._username, this._password);
+  User(this._username, this._password, {String user});
```
Comment (manishiitg) [inline lib/model/user.dart:8]: this should extend equatable class

File: lib/bloc/login/login_state.dart [hunk 14]
```diff
@@ -0,0 +1,22 @@
+import 'package:meta/meta.dart';
+import 'package:equatable/equatable.dart';
+
+abstract class LoginState extends Equatable {
+  @override
+  List<Object> get props => [];
+}
+
```
Comment (manishiitg) [inline lib/bloc/login/login_state.dart:8]: there is an login state and authentication state? not clear?

File: lib/model/user.dart (hunk 13 above)
Comment (flutterexcel) [reply lib/model/user.dart]: @manishiitg did it with equatable

File: lib/services/authentication_services.dart (hunk 11 above)
Comment (flutterexcel) [reply lib/services/authentication_services.dart]: @manishiitg this's done as well

File: lib/bloc/authentication/authentication_bloc.dart [hunk 15]
```diff
@@ -0,0 +1,60 @@
+import 'package:bloc/bloc.dart';
+
+import 'authentication_event.dart';
+import 'authentication_state.dart';
+import '../../services/authentication_services.dart';
+
+class AuthenticationBloc
+    extends Bloc<AuthenticationEvent, AuthenticationState> {
+  final AuthenticationService _authenticationService;
+
```
Comment (manishiitg) [inline lib/bloc/authentication/authentication_bloc.dart:10]: ok. i dont think we need a seperate authentication bloc. this is over compliecation.

we can just add another event "CHECK_AUTHENTICATED" in the login block itself. 

as this is just to check.

Comment (manishiitg) [reply lib/bloc/authentication/authentication_bloc.dart:10]: so we can remove this whole bloc

File: lib/LoginScreen.dart (hunk 5 above)
Comment (flutterexcel) [reply lib/LoginScreen.dart]: this is used for color and font

File: lib/screens/loginPage.dart (hunk 6 above)
Comment (flutterexcel) [reply lib/screens/loginPage.dart]: this is resolved as well

File: lib/screens/login_screen.dart (hunk 12 above)
Comment (manishiitg) [reply lib/screens/login_screen.dart:2]: ok. so this structure is wrong.
as u have kept both of the file in screens/ folder. this means two actual different screens for the user.

we should have /screens/login.dart
and then can we have /widgets/login/.....  etc inside this all widgets for login

File: lib/services/storage_service.dart (hunk 7 above)
Comment (manishiitg) [reply lib/services/storage_service.dart:25]: can u take action on this

Comment (flutterexcel) [reply lib/services/storage_service.dart:25]: it's pending still, we see to modify it

File: lib/screens/login_screen.dart (hunk 12 above)
Comment (flutterexcel) [reply lib/screens/login_screen.dart:2]: ok will change

File: lib/services/storage_service.dart (hunk 7 above)
Comment (flutterexcel) [reply lib/services/storage_service.dart:25]: @manishiitg modified it

File: lib/screens/login_screen.dart (hunk 12 above)
Comment (flutterexcel) [reply lib/screens/login_screen.dart:2]: @manishiitg done

File: lib/bloc/login/login_bloc.dart [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    try {
+      final user = await _authenticationService.signInWithEmailAndPassword(
+          event.email, event.password);
+
+      if (user != null) {
+        // push new login event
+
+        this.add(UserLogIn(user: user));
+        yield LoginSuccess();
+        yield LoginInitial();
```
Comment (manishiitg) [inline lib/bloc/login/login_bloc.dart:73]: what does Login Initial Do?

File: lib/main.dart [hunk 17]
```diff
@@ -1,28 +1,51 @@
...
+import 'services/authentication_services.dart';
 
-void main() {
-  runApp(HrApp());
-}
+void main() => runApp(
+
+        // Injects the Authentication service
+        RepositoryProvider<AuthenticationService>(
+      create: (context) {
```
Comment (manishiitg) [inline lib/main.dart:12]: format this

File: lib/services/storage_service.dart (hunk 7 above)
Comment (manishiitg) [reply lib/services/storage_service.dart:61]: this is not done. discuss with me

File: lib/bloc/login/login_bloc.dart (hunk 16 above)
Comment (flutterexcel) [reply lib/bloc/login/login_bloc.dart:73]: Not @manishiitg  needed, removed it.

**Pull request merged**